pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,438 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package featurestates
+
+import (
+	"context"
+	"reflect"
+	"strconv"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/internalapis"
+	featurestatesv1alpha1 "sigs.k8s.io/vsphere-csi-driver/pkg/internalapis/featurestates/v1alpha1"
+	k8s "sigs.k8s.io/vsphere-csi-driver/pkg/kubernetes"
+)
+
+const (
+	// CRDName represent the name of cnscsisvfeaturestate CRD
+	CRDName = "cnscsisvfeaturestates.cns.vmware.com"
+	// CRDGroupName represent the group of cnscsisvfeaturestate CRD
+	CRDGroupName = "cns.vmware.com"
+	// CRDSingular represent the singular name of cnscsisvfeaturestate CRD
+	CRDSingular = "cnscsisvfeaturestate"
+	// CRDPlural represent the plural name of cnscsisvfeaturestates CRD
+	CRDPlural = "cnscsisvfeaturestates"
+	// WorkLoadNamespaceLabelKey is the label key found on the workload namespace in the supervisor k8s cluster
+	WorkLoadNamespaceLabelKey = "vSphereClusterID"
+	// SVFeatureStateCRName to be used for CR in workload namespaces
+	SVFeatureStateCRName = "svfeaturestates"
+	// crUpdateRetryInterval is the interval at which pending CR update/create tasks are executed
+	crUpdateRetryInterval = 30 * time.Second
+)
+
+// pendingCRUpdates holds latest states of the namespaces to push CR update, latest feature states and the lock to operate
+// on namespaceUpdateMap and latestFeatureStates
+type pendingCRUpdates struct {
+	// lock to acquire before updating namespaceUpdateMap and latestFeatureStates
+	lock *sync.RWMutex
+	// holds namespace name and status to update CR in the namespace
+	// map value true, means update is pending, false means no update is required.
+	namespaceUpdateMap map[string]bool
+	// latest featureStates
+	latestFeatureStates []featurestatesv1alpha1.FeatureState
```

Comment:
Why is this required? Why cant processPendingCRUpdates query the latest FSS when it runs?

Reply:
latestFeatureStates is loaded with the latest feature state when service starts, and we keep this up to date with config-map informer by watching on FSS config-map using informers.  This helps avoid unnecessary calls to k8s to query the latest FSS.